src/serac/physics/thermal_conduction_functional.hpp

```diff
@@ -119,23 +120,47 @@ class ThermalConductionFunctional : public BasePhysics {
    *
    * @param[in] options The system linear and nonlinear solver and timestepping parameters
    * @param[in] name An optional name for the physics module instance
+   * @param[in] parameter_states The optional array of finite element states represetnting user-defined parameters to be
+   * used by an underlying material model or load
    */
-  ThermalConductionFunctional(const SolverOptions& options, const std::string& name = "")
-      : BasePhysics(1, order),
+  ThermalConductionFunctional(
+      const SolverOptions& options, const std::string& name = {},
+      std::array<std::reference_wrapper<FiniteElementState>, sizeof...(parameter_space)> parameter_states = {})
+      : BasePhysics(2, order),
         temperature_(
             StateManager::newState(FiniteElementState::Options{.order      = order,
                                                                .vector_dim = 1,
                                                                .ordering   = mfem::Ordering::byNODES,
                                                                .name       = detail::addPrefix(name, "temperature")})),
-        M_functional_(&temperature_.space(), {&temperature_.space()}),
-        K_functional_(&temperature_.space(), {&temperature_.space()}),
+        adjoint_temperature_(StateManager::newState(
+            FiniteElementState::Options{.order      = order,
+                                        .vector_dim = 1,
+                                        .ordering   = mfem::Ordering::byNODES,
+                                        .name       = detail::addPrefix(name, "adjoint_temperature")})),
+        parameter_states_(parameter_states),
         residual_(temperature_.space().TrueVSize()),
         ode_(temperature_.space().TrueVSize(), {.u = u_, .dt = dt_, .du_dt = previous_, .previous_dt = previous_dt_},
              nonlin_solver_, bcs_)
   {
     SLIC_ERROR_ROOT_IF(mesh_.Dimension() != dim,
                        axom::fmt::format("Compile time dimension and runtime mesh dimension mismatch"));
 
+    // Create a pack of the primal field and parameter finite element spaces
+    std::array<mfem::ParFiniteElementSpace*, sizeof...(parameter_space) + 1> trial_spaces;
+    trial_spaces[0] = &temperature_.space();
+
+    functional_call_args_.emplace_back(temperature_.trueVec());
+
+    for (long unsigned int i = 0; i < sizeof...(parameter_space); ++i) {
```

Comment:
`egrep -r "long  *unsigned  *int" serac` doesn't show any other uses of `long unsigned int` in the serac source. It seems that `std::size_t` is more common in cases like this. Of course, `unsigned char` is probably big enough to handle counting up to `sizeof...(parameter_space)`, but it's troublesome knowing that `sizeof...()` returns a `std::size_t`.

Reply:
Yeah, I'll do a pass to clean this up.